README.md
# HeavyMetal TTY Control Program

HeavyMetal is a perl based software with a Tk/Tcl based GUI, which allows your PC (be it Windows or Linux) to interact with a teleprinter (TTY) over the good old RS-232 serial port.

The software allows you to access different internet services, as well as be accessed by different types of clients. We could describe HeavyMetal as a kind of chat/messaging server, with support for many protocols, and above all the ability to interact with these old steel monsters we lovely call TTYs. One basic problem we face with TTYs is the need to feed them with suitable data, as well as the need to impress those who come home showing the top tech from the early XXth century still interacting with all the top tech of the XXIst century...

HeavyMetal was developed by Bill Buzbee from 2001 to 2006. You can access his webpage here, where you can get version 1.27

Between May and November 2010, I took the task of entirely rewriting the code into a more expandable codebase, while adding support for sessions and many new cool features. At the same time I kept the GUI more or less as it was.

In 2012 I was forced to give another jump into v3.1 due to changes in the ActivePerl distribution (they dropped the Tk extension I was using), so as I had to rewrite pretty much the whole GUI into a new module (Tkx). I obviously took the chance to make many changes that were in the ToDo list.

Project homepage: http://www.albinarrate.com/heavymetal.html

## HeavyMetal v3.1 in Linux

The following is a step by step list to get heavyMetal working on linux.
You may need to adapt some steps to best fit your needs.

1- Get the Fedora installed (Using VMWare image in my case)
	In your case you almost surely will not need to do this ;)

2- Allow root acces by modifying pam
	Files: /etc/pam.d/gdm and /etc/pam.d/gdm-password
	Comment out line: auth required pam_succeed_if.so user != root quiet
	Instructions: http://fedoraproject.org/wiki/Enabling_Root_User_For_GNOME_Display_Manager
	In your case you almost surely will not need to do this ;)

3- Login as root

4- Download and install latest ActivePerl community edition as a package
	http://www.activestate.com/activeperl/downloads
	In my case ActivePerl-5.14.2.1402-i686-linux-glibc-2.3.6-295342.tar.gz
	
5- Extract it with the archive manager

6- Follow instructions in readme.txt from the extracted folder:
	Open terminal
	cd into the extracted dir
	Then run: sh install.sh
	Install it into the default /opt/ActivePerl-blah
	
7- Edit /home/root/.bash_profile
	Added the following lines:
		export PATH=$HOME/bin:/opt/ActivePerl-5.14/bin:$PATH
		export PERL5LIB=/opt/ActivePerl-5.14/lib:/opt/ActivePerl-5.14/site/lib
	Logout and login again (apparently this was needed to reload the ENV vars?!)
	
8- Install the following packages with the OS packages tool (System Tools / Add/Remove Software):
	X.Org X11 libXss runtime library
	command line clipboard grabber (xclip)
	
9- With the Activeperl ppm, install the packages
	cd /opt/ActivePerl-5.14/bin
	ppm-shell 
		install Device-SerialPort
		install Clipboard
		install Crypt-SSLeay
		install Finance-YahooQuote
		install XML-RSS-Parser
		
	Or equivalent actions with the GUI based ppm
	
10- DONE! Go to the heavymetal dir, and do:
	perl heavymetal.pl
	
	According to your setup, you might need to provide an absolute path to perl, or to edit the first line of heavymetal.pl to indicate to the shell, where is perl.

## HeavyMetal v3.1 in Windows

The following is a step by step list to get heavyMetal working on Windows.
You may need to adapt some steps to best fit your needs.
Please note that is you are using an HeavyMetal version with ActivePerl 
already packaged, you do not need to follow these instructions. 

1- Download and install latest ActivePerl community edition 32 bits x86 
	Do not use 64 bits!
	http://www.activestate.com/activeperl/downloads
	In my case ActivePerl-5.14.2.1402-MSWin32-x86-295342.msi
	
2- Once installed, execute the Perl Package Manager you can find in the 
   Start menu or use the ppm-shell command

3- Install the following components:
	Tkx
	Tkx-Scrolled
	Win32-API
	Clipboard
	Crypt-SSLeay
	Finance-YahooQuote
	HTML-TreeBuilder
	Weather-Google
	XML-RSS-Parser

	Remember that some of these may already be installed in your perl 
	setup while some other module not mentioned here may be missing 
	in your setup.
	
	During HM initialization, you will see a list of perl modules being
	loaded. If some one fails it will be notified there. Also keep in 
	mind that most modules are optional and they are reported as 
	OPTIONAL. Hence if a module does not load, HM will initialize 
	properly, but the features that require the failed module will 
	simply not work.
	
4- DONE! Go to the heavymetal dir, and do:
	perl heavymetal.pl
	
	You may need to add the perl/bin folder to your path environment 
	variable.
	On initialization the list of modules will come up and if there 
	is any error it will be mentioned. You might need to install more
	packages according to your environment.
	Most modules are optional, and if they do not load, whatever feature
	that relies on them will be disabled.
	That means that even with some modules missing, you may get the GUI working.


##Version History

### HeavyMetal v3.0 (2010-06-25)
* Session based with independent buffers
* Support for Telnet clients and servers
* Ability to pipe a telnet connection between two HMs so you can chat TTY-to-TTY over Internet
* MSN client allows to send and receive messages and commands via MSN
* Eyeball labels in punched tape
* Expandable support for different encodings
* Whole new way to setup configurations, easily expandable
* Reorganized commands so new commands can be easily added
* Fixed AP News & Today in History
* Fixed SMTP Auth
* Fixed NOAA FTP url used for Weather forecasts

### HeavyMetal v3.1.000 (2012-02-28)
* Rewrote the GUI, now uses tabs and configs are available right at the GUI. You even have icons!
* Support for ConsoleOnly mode. With it you can run HM as a daemon and access it via Telnet.
* Support for multiple serial ports (max 2 for now, but easily expandable) with totally independent configurations.
* RSS News feeds now allow to add any source for the news summary
* Full news supported for selected sources. Right now Reuters and BBC. We can easily add more on demand.
* Custom commands, now you can write your own command, and configure it right from the GUI
* Loop echo detection (it automatically checks for echo and enables loop echo suppression with just one click)
* Favorites for everything, now you can easily setup at the GUI your favorite Weather cities, telnet hosts, news feeds, etc. They will simply show up in the menu.
* A system to autoupdate the software with a single click
* Fix to the weather FTP from NOAA, now the cities are dynamically loaded instead of having a static list (which was outdated)
* A plethora of bugfixes. Now it has been tested in Linux (Fedora using ActivePerl)

### HeavyMetal v3.1.001 (2012-03-10)
* New RSS tab to handle many many RSS feeds
* Many Favorites added to menus
* Main weather provider switched to WWO
* Added special RSS feed for historic cables
* Added WEATHER METAR support in various formats and with subcommand
* Added a progress bar for measuring the buffer in a nice graphical way

### HeavyMetal v3.1.002 (2012-03-25)
* New built-in CRON for scheduled tasks
* Twitter feeds support via RSS
* METAR HISTORIC format, just like the weather teletypes from the 1940s
* NOAA CLIMATE access added
* NOAA menues are now cached
* Double click on a feed in the RSS tab now shows a summary for that feed.
* $WEB command for very basic HTML text-browsing
* New built-in BANNER command, in replacement of the external command used in previous versions

### HeavyMetal v3.1.003 (2012-04-08)
* Bugfixes in several places, including fixing TTY2 session
* DTR & RTS line control on the serial port (i.e. for John Nagle board)
* $MOTOR command and auto turn on/off with safety delays included
* Easy one click connect from menu to another user of HMNet user
* Weather temperatures are now Celsius/Farenheit configurable
* Option to send report for bugs using the Help menu, one single click

### HeavyMetal v3.1.004 (2013-03-25)
* Support for External Codes (tables in ini files)
* External Custom Commands (custom.pl)
* Remote Control.

USB SiLabs CP2102/ReleaseNotes.txt
CP210x Baud Rate Configuration Utility v2.2b Release Notes
Copyright (C) 2010 Silicon Laboratories, Inc.

This release contains the following components:

        * CP210xBaudRateAliasConfig.exe (2.2)
        * CP210xBaudRateAliasConfig_SRC.zip
        * CP210xManufacturing.dll (4.1)
        * CP210xManufacturing.lib
        * CP210xManufacturingDLL.h
        * ReleaseNotes.txt (this file)

KNOWN ISSUES AND LIMITATIONS
----------------------------

	1.) Only CP2101, CP2102, CP2103 and CP2104 devices are supported by 
	    this utility. 

	2.) CP210xManufacturing.dll must be in the same directory as
	    CP210xBaudRateAliasConfig.exe.

Release Dates
-------------

	CP210x Baud Rate Configuration Utility v2.2c - July 26, 2010


REVISION HISTORY
-----------------

version 2.2c
	The CP210xManufacturing.DLL has been updated to version 4.1

version 2.2b
	The CP210xManufacturing.DLL has been updated to version 3.5 for
	CP2104 support

version 2.2a
	The CP210xManufacturing.DLL has been updated to version 3.4

version 2.2
	Fixed possible divide by zero errors in baud rate calculation
	The CP210xManufacturing.DLL has been updated to version 3.3

version 2.1
	The CP210xManufacturing.DLL has been updated to version 3.0 to add support
	for the 5.x driver.

version 2.0
	Changed GetNumDevices to not return an error when there are 0 devices connected.
	Fixed the "Locked Byte always 1" bug.

version 1.2
	CP210xManufacturing DLL CP210x_GetProductString, CP210x_SetProductString,
	CP210x_GetSerialNumber and CP210x_SetSerialNumber updated so that the UNICODE
	options work according to documentation, length is returned as number of characters
	without the NULL character.

	CP210xManufacturing DLL PORT_CONFIG structure updated according to documentation.

version 1.1
	The CP210x Manufacturing DLL has been updated. See AN144 and 
	AN144SW for further details.

Version 1.0
	Initial Release



lib.custom/Finance/YahooQuote.pm
# perl -w
#
#    Copyright (C) 1998-2002, Dj Padzensky <djpadz@padz.net>
#    Copyright (C) 2002-2010  Dirk Eddelbuettel <edd@debian.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# $Id: YahooQuote.pm,v 1.11 2010/03/27 00:44:10 edd Exp $

package Finance::YahooQuote;
require 5.005;

require Exporter;
use strict;
use vars qw($VERSION @EXPORT @ISA 
	    $QURL $QURLbase $QURLformat $QURLextended $QURLrealtime $QURLend
	    $TIMEOUT $PROXY $PROXYUSER $PROXYPASSWD);

use HTTP::Request::Common;
use Text::ParseWords;

$VERSION = '0.24';

## these variables govern what type of quote the modules is retrieving
$QURLbase = "http://download.finance.yahoo.com/d/quotes.csvr?e=.csv&f=";
$QURLformat = "snl1d1t1c1p2va2bapomwerr1dyj1x";	# default up to 0.19
$QURLextended = "s7t8e7e8e9r6r7r5b4p6p5j4m3m4";	# new in 0.20
$QURLrealtime = "b2b3k2k1c6m2j3"; # also new in 0.20
$QURLend = "&s=";
$QURL = $QURLbase . $QURLformat . $QURLend; # define old format as default

@ISA = qw(Exporter);
@EXPORT = qw(&getquote &getonequote &getcustomquote 
	     &setQueryString &useExtendedQueryFormat 
	     &useRealtimeQueryFormat);
undef $TIMEOUT;

## simple function to switch to extended format
sub useExtendedQueryFormat {
  $QURLformat .= $QURLextended;
}

## simple function to append real-time format
sub useRealtimeQueryFormat {
  $QURLformat .= $QURLrealtime;
}

## allow module user to define a new query string
sub setQueryString {
  my $format = shift;
  $QURLformat = $format;
}

## Common function for reading the yahoo site and retrieving and
## parsing the data returned
## The format used is supplied as a second argument, the default function
## getquoye() below uses the default format $QURLformat, and either one of
## the two preceding functions can be used to select the extended format
sub readYahoo {

  my @symbols	= @{(shift)};
  my $format	= shift;

  my @qr;
  my $ua = RequestAgent->new;
  $ua->env_proxy;		# proxy settings from *_proxy env. variables.
  $ua->proxy('http', $PROXY) if defined $PROXY;
  $ua->timeout($TIMEOUT) if defined $TIMEOUT;

  #	Loop over the list of symbols, grabbing 199 symbols at a 
  #	time, since yahoo only lets you get 200 per request
  while ($#symbols > -1) {
    my @b = $#symbols >= 199 	? splice(@symbols,0,199) 
      : splice(@symbols,0,$#symbols+1);

    my $url = "$QURLbase${format}$QURLend".join('+',@b);
    foreach (split('\015?\012',$ua->request(GET $url)->content)) {
      my @q = quotewords(',',0,$_);
      push(@qr,[@q]);
    }
  }

  return \@qr;
}

my %fields = (
	      'Symbol'				=>  's',	# old default
	      'Name'				=>  'n',	# old default
	      'Last Trade (With Time)'		=>  'l',
	      'Last Trade (Price Only)'		=>  'l1',	# old default
	      'Last Trade Date'			=>  'd1',	# old default
	      'Last Trade Time'			=>  't1',	# old default
	      'Last Trade Size'			=>  'k3',
	      'Change and Percent Change'	=>  'c',
	      'Change'				=>  'c1',	# old default
	      'Change in Percent'		=>  'p2',	# old default
	      'Ticker Trend'			=>  't7',
	      'Volume'				=>  'v',	# old default
	      'Average Daily Volume'		=>  'a2',	# old default
	      'More Info'			=>  'i',
	      'Trade Links'			=>  't6',
	      'Bid'				=>  'b',	# old default
	      'Bid Size'			=>  'b6',
	      'Ask'				=>  'a',	# old default
	      'Ask Size'			=>  'a5',
	      'Previous Close'			=>  'p',	# old default
	      'Open'				=>  'o',	# old default
	      "Day's Range"			=>  'm',	# old default
	      '52-week Range'			=>  'w',	# old default
	      'Change From 52-wk Low'		=>  'j5',
	      'Pct Chg From 52-wk Low'		=>  'j6',
	      'Change From 52-wk High'		=>  'k4',
	      'Pct Chg From 52-wk High'		=>  'k5',
	      'Earnings/Share'			=>  'e',	# old default
	      'P/E Ratio'			=>  'r',	# old default
	      'Short Ratio'			=>  's7',
	      'Dividend Pay Date'		=>  'r1',	# old default
	      'Ex-Dividend Date'		=>  'q',
	      'Dividend/Share'			=>  'd',	# old default
	      'Dividend Yield'			=>  'y',	# old default
	      'Float Shares'			=>  'f6',
	      'Market Capitalization'		=>  'j1',	# old default
	      '1yr Target Price'		=>  't8',
	      'EPS Est. Current Yr'		=>  'e7',
	      'EPS Est. Next Year'		=>  'e8',
	      'EPS Est. Next Quarter'		=>  'e9',
	      'Price/EPS Est. Current Yr'	=>  'r6',
	      'Price/EPS Est. Next Yr'		=>  'r7',
	      'PEG Ratio'			=>  'r5',
	      'Book Value'			=>  'b4',
	      'Price/Book'			=>  'p6',
	      'Price/Sales'			=>  'p5',
	      'EBITDA'				=>  'j4',
	      '50-day Moving Avg'		=>  'm3',
	      'Change From 50-day Moving Avg'	=>  'm7',
	      'Pct Chg From 50-day Moving Avg'	=>  'm8',
	      '200-day Moving Avg'		=>  'm4',
	      'Change From 200-day Moving Avg'	=>  'm5',
	      'Pct Chg From 200-day Moving Avg'	=>  'm6',
	      'Shares Owned'			=>  's1',
	      'Price Paid'			=>  'p1',
	      'Commission'			=>  'c3',
	      'Holdings Value'			=>  'v1',
	      "Day's Value Change"		=>  'w1',
	      'Holdings Gain Percent'		=>  'g1',
	      'Holdings Gain'			=>  'g4',
	      'Trade Date'			=>  'd2',
	      'Annualized Gain'			=>  'g3',
	      'High Limit'			=>  'l2',
	      'Low Limit'			=>  'l3',
	      'Notes'				=>  'n4',
	      'Last Trade (Real-time) with Time'=>  'k1',
	      'Bid (Real-time)'			=>  'b3',
	      'Ask (Real-time)'			=>  'b2',
	      'Change Percent (Real-time)'	=>  'k2',
	      'Change (Real-time)'		=>  'c6',
	      'Holdings Value (Real-time)'	=>  'v7',
	      "Day's Value Change (Real-time)"	=>  'w4',
	      'Holdings Gain Pct (Real-time)'	=>  'g5',
	      'Holdings Gain (Real-time)'	=>  'g6',
	      "Day's Range (Real-time)"		=>  'm2',
	      'Market Cap (Real-time)'		=>  'j3',
	      'P/E (Real-time)'			=>  'r2',
	      'After Hours Change (Real-time)'	=>  'c8',
	      'Order Book (Real-time)'		=>  'i5',
	      'Stock Exchange'			=>  'x'		# old default
	     );

#	Let the user define which colums to retrive from yahoo
#	
sub getcustomquote {
  my $symbols = shift;
  my $columns = shift;

  my $format = join('',map {$fields{$_}} @{$columns});

  my $qr = readYahoo($symbols,$format);
  return wantarray() ? @$qr : $qr;
}

# get quotes for all symbols in array

sub getquote {
  my @symbols = @_;

  my $format = $QURLformat;	## Old default from variable
  my $qr = readYahoo(\@symbols,$format);
  return wantarray() ? @$qr : $qr;
}

# Input: A single stock symbol
# Output: An array, containing the list elements mentioned above.

sub getonequote {
    my @x;
    @x = &getquote($_[0]);
    return wantarray() ? @{$x[0]} : \@{$x[0]} if @x;
}

BEGIN {				# Local variant of LWP::UserAgent that 
  use LWP;			# checks for user/password if document 
  package RequestAgent;		# this code taken from lwp-request, see
  no strict 'vars';		# the various LWP manual pages
  @ISA = qw(LWP::UserAgent);

  sub new { 
    my $self = LWP::UserAgent::new(@_);
    $self->agent("Finance-YahooQuote/0.18");
    $self;
  }

  sub get_basic_credentials {
    my $self = @_;
    if (defined($PROXYUSER) and defined($PROXYPASSWD) and
 	$PROXYUSER ne "" and $PROXYPASSWD ne "") {
      return ($PROXYUSER, $PROXYPASSWD);
    } else {
      return (undef, undef)
    }
  }
}

1;

__END__

=head1 NAME

Finance::YahooQuote - Get stock quotes from Yahoo! Finance

=head1 SYNOPSIS

  use Finance::YahooQuote;
  # setting TIMEOUT and PROXY is optional
  $Finance::YahooQuote::TIMEOUT = 60;
  $Finance::YahooQuote::PROXY = "http://some.where.net:8080";
  @quote = getonequote $symbol;	# Get a quote for a single symbol
  @quotes = getquote @symbols;	# Get quotes for a bunch of symbols
  useExtendedQueryFormat();     # switch to extended query format
  useRealtimeQueryFormat();     # switch to real-time query format
  @quotes = getquote @symbols;	# Get quotes for a bunch of symbols
  @quotes = getcustomquote(["DELL","IBM"], # using custom format
			   ["Name","Book Value"]); # note array refs

=head1 DESCRIPTION

This module gets stock quotes from Yahoo! Finance.  The B<getonequote>
function will return a quote for a single stock symbol, while the
B<getquote> function will return a quote for each of the stock symbols
passed to it.  B<getcustomquote> allows to specify a format other than
the default to take advantage of the extended range of available information.

The download operation is efficient: only one request is made even if
several symbols are requested at once. The return value of
B<getonequote> is an array, with the following elements:

    0 Symbol
    1 Company Name
    2 Last Price
    3 Last Trade Date
    4 Last Trade Time
    5 Change
    6 Percent Change
    7 Volume
    8 Average Daily Vol
    9 Bid
    10 Ask
    11 Previous Close
    12 Today's Open
    13 Day's Range
    14 52-Week Range
    15 Earnings per Share
    16 P/E Ratio
    17 Dividend Pay Date
    18 Dividend per Share
    19 Dividend Yield
    20 Market Capitalization
    21 Stock Exchange

If the extended format has been selected, the following fields are also
retrieved:

    22 Short ratio
    23 1yr Target Price
    24 EPS Est. Current Yr
    25 EPS Est. Next Year
    26 EPS Est. Next Quarter
    27 Price/EPS Est. Current Yr
    28 Price/EPS Est. Next Yr
    29 PEG Ratio
    30 Book Value
    31 Price/Book
    32 Price/Sales
    33 EBITDA
    34 50-day Moving Avg
    35 200-day Moving Avg

If the real-time format has been selected, the following fields are also
retrieved:

    36 Ask (real-time)
    37 Bid (real-time)
    38 Change in Percent (real-time)
    39 Last trade with time (real-time)
    40 Change (real-time)
    41 Day range (real-time)
    42 Market-cap (real-time)

The B<getquote> function returns an array of pointers to arrays with
the above structure.

The B<getonequote> function returns just one quote, rather than an array. It returns a simple array of values for the given symbol.

The B<setQueryString> permits to supply a new query string that will
be used for subsequent data requests.

The B<useExtendedQueryFormat> and B<useRealtimeQueryFormat> are
simpler interfaces which append symbols to the default quote string,
as detailed above.

The B<getcustomquote> returns an array of quotes corresponding to
values for the symbols supplied in the first array reference, and the
custom fields supplied in the second array reference. Here the custom
fields correspond to the 'named' fields of the list below.

Beyond stock quotes, B<Finance::YahooQuote> can also obtain quotes for
currencies (from the Philadephia exchange -- however Yahoo! appears to
have stopped to support the currency symbols in a reliable manner), US
mutual funds, options on US stocks, several precious metals and quite
possibly more; see the Yahoo! Finance website for full
information. B<Finance::YahooQuote> can be used for stocks from the
USA, Canada, various European exchanges, various Asian exchanges
(Singapore, Taiwan, HongKong, Kuala Lumpur, ...) Australia and New
Zealand. It should work for other markets supported by Yahoo.

You may optionally override the default LWP timeout of 180 seconds by setting
$Finance::YahooQuote::TIMEOUT to your preferred value.

You may also provide a proxy (for the required http connection) by using
the variable $Finance::YahooQuote::PROXY. Furthermore, authentication-based 
proxies can be used by setting the proxy user and password via the variables
$Finance::YahooQuote::PROXYUSER and $Finance::YahooQuote::PROXYPASSWD.

Two example scripts are provided to help with the mapping a stock
symbols as well as with Yahoo! Finance server codes. The regression
tests scripts in the B<t/> subdirectory of the source distribution
also contain simple examples.

=head2 The available custom fields

The following list contains all the available data fields at Yahoo!
along with the corresponding format string entry:

      Symbol				s
      Name				n
      Last Trade (With Time)		l
      Last Trade (Price Only)		l1
      Last Trade Date			d1
      Last Trade Time			t1
      Last Trade Size			k3
      Change and Percent Change		c
      Change				c1
      Change in Percent			p2
      Ticker Trend			t7
      Volume				v
      Average Daily Volume		a2
      More Info				i
      Trade Links			t6
      Bid				b
      Bid Size				b6
      Ask				a
      Ask Size				a5
      Previous Close			p
      Open				o
      Day's Range			m
      52-week Range			w
      Change From 52-wk Low		j5
      Pct Chg From 52-wk Low		j6
      Change From 52-wk High		k4
      Pct Chg From 52-wk High		k5
      Earnings/Share			e
      P/E Ratio				r
      Short Ratio			s7
      Dividend Pay Date			r1
      Ex-Dividend Date			q
      Dividend/Share			d
      Dividend Yield			y
      Float Shares			f6
      Market Capitalization		j1
      1yr Target Price			t8
      EPS Est. Current Yr		e7
      EPS Est. Next Year		e8
      EPS Est. Next Quarter		e9
      Price/EPS Est. Current Yr		r6
      Price/EPS Est. Next Yr		r7
      PEG Ratio				r5
      Book Value			b4
      Price/Book			p6
      Price/Sales			p5
      EBITDA				j4
      50-day Moving Avg			m3
      Change From 50-day Moving Avg	m7
      Pct Chg From 50-day Moving Avg	m8
      200-day Moving Avg		m4
      Change From 200-day Moving Avg	m5
      Pct Chg From 200-day Moving Avg	m6
      Shares Owned			s1
      Price Paid			p1
      Commission			c3
      Holdings Value			v1
      Day's Value Change		w1,
      Holdings Gain Percent		g1
      Holdings Gain			g4
      Trade Date			d2
      Annualized Gain			g3
      High Limit			l2
      Low Limit				l3
      Notes				n4
      Last Trade (Real-time) with Time	k1
      Bid (Real-time)			b3
      Ask (Real-time)			b2
      Change Percent (Real-time)	k2
      Change (Real-time)		c6
      Holdings Value (Real-time)	v7
      Day's Value Change (Real-time)	w4
      Holdings Gain Pct (Real-time)	g5
      Holdings Gain (Real-time)		g6
      Day's Range (Real-time)		m2
      Market Cap (Real-time)		j3
      P/E (Real-time)			r2
      After Hours Change (Real-time)	c8
      Order Book (Real-time)		i5
      Stock Exchange			x

=head1 FAQs

=head2 How can one figure out the format string?  

Provided a My Yahoo! (http://my.yahoo.com) account, go to the
following URL:

    http://edit.my.yahoo.com/config/edit_pfview?.vk=v1

Viewing the source of this page, you will come across the section that
defines the menus that let you select which elements go into a
particular view.  The <option> values are the strings that pick up
the information described in the menu item.  For example, Symbol
refers to the string "s" and name refers to the string "l".  Using
"sl" as the format string, we would get the symbol followed by the
name of the security.

The example script I<examine_server.sh> shows this in some more detail
and downloads example .csv files using B<GNU wget>.

=head2 What about different stock symbols for the same corporation?

This can be issue. For the first few years, Yahoo! Finance's servers 
appeared to be cover their respective local markets. E.g., the UK-based 
servers provided quotes for Europe, the Australian one for the Australia
and New Zealand and so on.  Hence, one needed to branch and bound code
and map symbols to their region's servers.

It now appears that this is no longer required, which is good news as it 
simplifies coding. However, some old symbols are no longer supported --
yet other, and supported, codes exist for the same company.  For example,
German stocks used to quoted in terms or their cusip-like 'WKN'. The
main server does not support these, but does support newer, acronym-based
symbols.  The example script examine_server.sh helps in finding the mapping
as e.g. from 555750.F to DTEGN.F for Deutsche Telekom. 

=head1 COPYRIGHT

Copyright 1998 - 2002 Dj Padzensky
Copyright 2002 - 2007 Dirk Eddelbuettel

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

The information that you obtain with this library may be copyrighted
by Yahoo! Inc., and is governed by their usage license.  See
http://www.yahoo.com/docs/info/gen_disclaimer.html for more
information.

=head1 AUTHOR

Dj Padzensky (C<djpadz@padz.net>), PadzNet, Inc., wrote the original
version. Dirk Eddelbuettel (C<edd@debian.org>) provided several
extensions based on DJ's original work and is the current maintainer.

=head1 SEE ALSO

The B<Finance::YahooQuote> home pages are found at
http://www.padz.net/~djpadz/YahooQuote/ and
http://dirk.eddelbuettel.com/code/yahooquote.html.

The B<smtm> (Show Me The Money) program uses Finance::YahooQuote for a
customisable stock/portfolio ticker and chart display, see 
http://dirk.eddelbuettel.com/code/smtm.html for more.  The B<beancounter>
program uses it to store quotes in a SQL database, see
http://dirk.eddelbuettel.com/code/beancounter.html.

=cut

